docs/src/index.md

```diff
@@ -0,0 +1,347 @@
+
+The `frule` is written:
+```julia
+function frule(::typeof(foo), args..., Δself, Δargs...; kwargs...)
+    ...
+    return y, ∂Y
+end
+```
+where again `y = foo(args; kwargs...)`,
+and `∂Y` is the result of propagating the derivative information forwards at that point.
+This propagation is call the pushforward.
+One could think of writing `∂Y = pushforward(Δself, Δargs)`, and often we will think of the `frule` as having the primal computation `y = foo(args...; kwargs...)`, and the push-forward `∂Y = pushforward(Δself, Δargs...)`
+
+
+!!! note "Why `rrule` returns a pullback but `frule` doesn't return a pushforward"
+    While `rrule` takes only the arguments to the original function (the primal arguments) and returns a function (the pullback) that operates with the derivative information, the `frule` does it all at once.
+    This is because the `frule` fuses the primal computation and the pushforward.
+    This is an optimization that allows `frule`s to contain single large operations that perform both the primal computation and the pushforward at the same time (for example solving an ODE).
+This operation is only possible in forward mode (where `frule` is used) because the derivative information needed by the pushforward available with the `frule` is invoked -- it is about the primal function's inputs.
+    In contrast, in reverse mode the derivative information needed by the pullback is about the primal function's output.
+    Thus the reverse mode returns the pullback function which the caller (usually an AD system) keeps hold of until derivative information about the output is available.
+
+
+## The propagators: pushforward and pullback
+
+
+!!! terminology "pushforward and pullback"
+
+    _Pushforward_ and _pullback_ are fancy words that the autodiff community recently adopted from Differential Geometry.
+    The are broadly in agreement with the use of [pullback](https://en.wikipedia.org/wiki/Pullback_(differential_geometry)) and [pushforward](https://en.wikipedia.org/wiki/Pushforward_(differential)) in differential geometry.
+    But any geometer will tell you these are the super-boring flat cases. Some will also frown at you.
+    They are also sometimes described in terms of the jacobian:
+    The _pushforward_ is _jacobian vector product_ (`jvp`), and _pullback_ is _jacobian transpose vector product_ (`j'vp`).
+    Other terms that may be used include for _pullback_ the **backpropagator**, and by analogy for _pushforward_ the **forwardpropagator**, thus these are the _propagators_.
+    These are also good names because effectively they propagate wiggles and wobbles through them, via the chain rule.
+    (the term **backpropagator** may originate with ["Lambda The Ultimate Backpropagator"](http://www-bcl.cs.may.ie/~barak/papers/toplas-reverse.pdf) by Pearlmutter and Siskind, 2008)
+
+### Core Idea
+
+#### Less formally
```

Comment:
Do we need both a formal and informal section?

Reply:
Idk. Out of scope for this PR.
I think this bit needs reworking